store/cafs/src/writeBufferToCafs.ts

```diff
@@ -87,14 +84,14 @@ function removeSuffix (filePath: string): string {
   return withoutSuffix
 }
 
-async function existsSame (filename: string, integrity: ssri.IntegrityLike) {
-  let existingFile: Stats | undefined
+function existsSame (filename: string, integrity: ssri.IntegrityLike) {
+  let existingFile: fs.Stats | undefined
   try {
-    existingFile = await fs.stat(filename)
+    existingFile = fs.statSync(filename)
```

Comment:
FYI, all currently supported versions of `fs.statSync` (16+) offer a `throwIfNoEntry: false` option which instead returns `undefined` if the file does not exist, to avoid the overhead of throwing and catching an exception.

I suspect for cafs collisions you may be losing a bit of performance due to each worker thread not being able to tell that the record was written by another thread. I'm not really sure on the best ways around that, though, since it involves some degree of cross-worker coordination to track which files have been written by the current process.

It probably wouldn't be too expensive to `readdir` the entire existing CAFS tree before the installer boots up and use that as the indicator of which files predate the current install and therefore might need integrity checking, vs. files that were just created by the current installation and can be assumed to be correct.

Reply:
Regarding reading the entire existing CAFS, we can do it as a later change as the performance in this PR is good enough at the moment. I've created an issue for it: https://github.com/pnpm/pnpm/issues/6921

Reply:
Done